### DISCUSSION TITLE
[neovim] Server connectivity issue

### DISCUSSION THREAD

**Discussion by keabarnes** [community/community#7578] [Copilot Conversations] [answered]
**Description**:

When running `:Copilot setup` I get the following error:
`Copilot: Server connectivity issue`

I can access `https://copilot-proxy.githubusercontent.com/_ping` from my browser manually.

`nvim --version` output:
```
NVIM v0.6.0-dev+208-g5c643dee7
Build type: Release
LuaJIT 2.1.0-beta3
Compiled by <redacted>

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/local/Cellar/neovim/HEAD-5c643de/share/nvim"

Run :checkhealth for more info

```

`:Copilot log` output:
```
Exception: 
  copilot#agent(-32603): Internal error: Error: GET https://copilot-proxy.githubusercontent.com/_ping timed out after 5000 ms 
  @ function copilot#Command[18]..4[1]..<SNR>54_NetworkStatusMessage[6]..copilot#HttpRequest[1]..copilot#agent#Call[2]..<SNR>110_AgentCall[5]..<SNR>110_RequestAwait, line 5

```

**Comment by keabarnes** [comment]
Still facing the same issue after:
- using Node 16 over 14
- updating the plugin to the latest version
- updating neovim to `v0.6.0-dev+572-g5d653a134`

**Comment by tpope** [comment]
Do you use a proxy server? Node does not use the system proxy configuration, so that could explain it.

**Reply by keabarnes** [reply]
I don't believe I do, I've created the following `main.js` file to test:
```javascript
const https = require("https");
const options = {
  hostname: "copilot-proxy.githubusercontent.com",
  port: 443,
  path: "/_ping",
  method: "GET",
};

const req = https.request(options, (res) => {
  console.log({ statusCode: res.statusCode });

  res.on("data", (d) => {
    process.stdout.write(d);
  });
});

req.on("error", (error) => {
  console.error(error);
});

req.end();
```

with result:
```json
{ "statusCode": 200 }
{"now":1636987315,"status":"ok"}
```

**Reply by keabarnes** [reply]
It's also not network related, as my other laptop doesn't have this issue on the same network

**Reply by tpope** [reply]
Can you test this variant?

```javascript
const {fetch} = require('fetch-h2');

async function pingTest() {
  const response = await fetch('https://copilot-proxy.githubusercontent.com/_ping', {
    timeout: 5000
  });

  console.log({statusCode: response.status})
  console.log(await response.text());
}

pingTest();
```

**Reply by keabarnes** [reply]
Sorry for the slow response, I've been traveling, the variant above exhibits the timeout behavior:

```bash
(node:36726) UnhandledPromiseRejectionWarning: Error: GET https://copilot-proxy.githubusercontent.com/_ping timed out after 5000 ms
    at timeoutError (./node_modules/fetch-h2/dist/lib/fetch-common.js:102:16)
    at Timeout._onTimeout (./node_modules/fetch-h2/dist/lib/fetch-common.js:125:28)
    at listOnTimeout (internal/timers.js:549:17)
    at processTimers (internal/timers.js:492:7)
```

Interesting that it seems to be something to the `fetch-h2` lib's specific implementation

**Reply by keabarnes** [reply]
Just though I'd make double sure, running below shows no proxies:
```bash
❯ sudo networksetup -getwebproxy Wi-Fi
Enabled: No
Server:
Port: 0
Authenticated Proxy Enabled: 0
```

If you have any other commands that I can run to rule out environment, I'd be happy to

**Reply by tpope** [reply]
`fetch-h2` is wrapping `http2`. Try this:

```js
const http2 = require('http2');
const fs = require('fs');
const client = http2.connect('https://copilot-proxy.githubusercontent.com');
client.on('error', (err) => console.error(err));

const req = client.request({ ':path': '/_ping' });

req.on('response', (headers, flags) => {
  for (const name in headers) {
    console.log(`${name}: ${headers[name]}`);
  }
});

req.setEncoding('utf8');
let data = '';
req.on('data', (chunk) => { data += chunk; });
req.on('end', () => {
  console.log(`\n${data}`);
  client.close();
});
req.end();
```

If I had to guess, I'd say you have a _transparent_ proxy between you and the Internet that doesn't support HTTP2. This won't show up in `networksetup -getwebproxy`.

**Reply by keabarnes** [reply]
Running the above script actually works, strangely enough, output below:

```http
status: 200
content-security-policy: default-src 'none'; sandbox
content-type: application/json
strict-transport-security: max-age=31536000
content-length: 33
date: Mon, 29 Nov 2021 06:54:36 GMT

{"now":1638168876,"status":"ok"}
```

**Reply by keabarnes** [reply]
Is there a reason `fetch-h2` was used over one of the other libraries?

**Reply by tpope** [reply]
Since it's literally a timeout error, I guess the obvious thing to try is omitting the `timeout` parameter from the `fetch-h2` version. And can you confirm it's actually taking 5 seconds before timing out? And the others succeed in less than 5 seconds?

**Reply by keabarnes** [reply]
Did some looking into how to detect transparent proxies, nothing conclusive, but got the suggestion to do a standard TCP connection check with the server - a transparent proxy will show a lower latency. I ran `nmap` against `copilot-proxy.githubusercontent.com`:

```
❯ nmap -sT copilot-proxy.githubusercontent.com -p 443
Starting Nmap 7.92 ( https://nmap.org ) at 2021-11-29 10:23 SAST
Nmap scan report for copilot-proxy.githubusercontent.com (51.138.70.97)
Host is up (0.15s latency).

PORT    STATE SERVICE
443/tcp open  https

Nmap done: 1 IP address (1 host up) scanned in 0.83 seconds
```

0.15s latency would suggest that it's not being terminated locally I believe.

**Reply by keabarnes** [reply]
> Since it's literally a timeout error, I guess the obvious thing to try is omitting the `timeout` parameter from the `fetch-h2` version. And can you confirm it's actually taking 5 seconds before timing out? And the others succeed in less than 5 seconds?

I've added some zeros to the timeout and it just sits forever, I'll try removing if there's some strange behaviour around including it.

**Reply by keabarnes** [reply]
> Since it's literally a timeout error, I guess the obvious thing to try is omitting the `timeout` parameter from the `fetch-h2` version. And can you confirm it's actually taking 5 seconds before timing out? And the others succeed in less than 5 seconds?

I've removed timeout and added some logging in the script as below:

```javascript
const { fetch } = require("fetch-h2");

async function pingTest() {
  console.log("Starting ping test.");
  try {
    const response = await fetch(
      "https://copilot-proxy.githubusercontent.com/_ping"
    );
    console.log("Received:");
    console.log({ statusCode: response.status });
    console.log(await response.text());
  } catch (e) {
    console.error({ e });
  }
}

pingTest();

```

Something even stranger is happening, the script exits with exit code `0` before getting to any of the logging:

```bash
❯ time node main.js
Starting ping test.
node main.js  0.14s user 0.03s system 6% cpu 2.631 total
❯ echo $?
0
```

**Reply by keabarnes** [reply]
`http2` version completes in less than a second:
```bash
❯ time node main2.js
:status: 200
content-security-policy: default-src 'none'; sandbox
content-type: application/json
strict-transport-security: max-age=31536000
content-length: 33
date: Mon, 29 Nov 2021 09:00:10 GMT

{"now":1638176410,"status":"ok"}

node main2.js  0.07s user 0.02s system 11% cpu 0.758 total
```

`fetch-h2` version times out after 5s if timeout param is included:
```bash
❯ time node main.js
Starting ping test.
{
  e: TimeoutError: GET https://copilot-proxy.githubusercontent.com/_ping timed out after 5000 ms
      at timeoutError (/Users/kb/software/adhoc/temp/copilot_testing/node_modules/fetch-h2/dist/lib/fetch-common.js:102:16)
      at Timeout._onTimeout (/Users/kb/software/adhoc/temp/copilot_testing/node_modules/fetch-h2/dist/lib/fetch-common.js:125:28)
      at listOnTimeout (internal/timers.js:549:17)
      at processTimers (internal/timers.js:492:7)
}
node main.js  0.15s user 0.03s system 3% cpu 5.677 total
```

**Reply by keabarnes** [reply]
@tpope Hope you had a good holiday break. Wondering if we can revive this? I've upgraded to the latest version and am still facing this issue.

**Reply by tpope** [reply]
I don't really have much to offer at this point other than try to figure out what's going wrong with `fetch-h2`. If I had an environment where I could reproduce the issue, I think my next step would be to point a packet sniffer at it and try to narrow down at what point in the connection life cycle it's timing out.

**Reply by keabarnes** [reply, marked as answer]
After updating, this seems to be resolved ✅
